test_agent/rust/src/utils.rs

```diff
@@ -0,0 +1,552 @@
+/*
+ * Copyright (c) 2024 General Motors GTO LLC
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * SPDX-FileType: SOURCE
+ * SPDX-FileCopyrightText: 2023 General Motors GTO LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+
+use serde_json::Value;
+use serde::{Deserialize, Deserializer};
+use up_rust::{
+    Data, UAttributes, UAuthority, UCode, UEntity, UMessage, UMessageType, UPayload,
+    UPayloadFormat, UPriority, UResource, UUri, UUID,
+};
+
+use protobuf::{MessageField, SpecialFields};
+
+pub fn convert_json_to_jsonstring<T: serde::Serialize>(value: &T) -> String {
+    serde_json::to_string(value).expect("Failed to convert to JSON string")
+}
+
+#[derive(Debug, Default)]
+pub struct WrapperUUri(pub UUri);
+impl<'de> Deserialize<'de> for WrapperUUri {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let value: Value = Deserialize::deserialize(deserializer)?;
+        //dbg!("WrapperUUri: {:?}", value.);
+
+        //update authority
+        let _authority_name = match value
```

Comment:
Could replace these `let foo = match bar` bits with:
```suggestion
        let authority_name = value
            .get("authority")
            .and_then(|authority| authority.get("name"))
            .and_then(|name| name.as_str())
            .unwrap_or("default");
```
to make it a little more idiomatic.

If you need a String instead of &str, you can chain .to_string() onto the end after the unwrap_or().

Reply:
corrected

Reply:
Mmmm, I'm not sure it was corrected. Can you take a look again?